File: app/src/main/java/com/rcdvl/marvel/characterdetails/CharacterDetailsFragment.java
package com.rcdvl.marvel.characterdetails;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.rcdvl.marvel.R;
import com.rcdvl.marvel.characterdetails.CharacterResourcePresenter.ResourceType;
import com.rcdvl.marvel.model.api.MarvelCharacter;
import com.rcdvl.marvel.model.api.MarvelUrl;

import org.parceler.Parcels;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * Created by renan on 16/06/16.
 */
public class CharacterDetailsFragment extends Fragment {

    private MarvelCharacter character;
    private Unbinder unbinder;
    @BindView(R.id.characterImage) ImageView characterImage;
    @BindView(R.id.characterName) TextView characterName;
    @BindView(R.id.characterDescription) TextView characterDescription;

    public static CharacterDetailsFragment newInstance(MarvelCharacter character) {
        Bundle args = new Bundle();
        args.putParcelable(CharacterDetailsActivity.EXTRA_CHARACTER, Parcels.wrap(character));

        CharacterDetailsFragment fragment = new CharacterDetailsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        character = Parcels.unwrap(getArguments().getParcelable(CharacterDetailsActivity.EXTRA_CHARACTER));
        return inflater.inflate(R.layout.fragment_character_details, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        unbinder = ButterKnife.bind(this, view);

        Glide.with(this).load(character.thumbnail.getImageUrl()).centerCrop().into(characterImage);

        characterName.setText(character.name);

        if (TextUtils.isEmpty(character.description)) {
            view.findViewById(R.id.descriptionContainer).setVisibility(View.GONE);
        } else {
            characterDescription.setText(character.description);
        }


        if (savedInstanceState == null) {
            CharacterResourceFragment fragment = CharacterResourceFragment.newInstance(character, ResourceType.COMICS);
            getChildFragmentManager().beginTransaction().replace(R.id.comicsListContainer, fragment).commit();

            fragment = CharacterResourceFragment.newInstance(character, ResourceType.SERIES);
            getChildFragmentManager().beginTransaction().replace(R.id.seriesListContainer, fragment).commit();

            fragment = CharacterResourceFragment.newInstance(character, ResourceType.STORIES);
            getChildFragmentManager().beginTransaction().replace(R.id.storiesListContainer, fragment).commit();

            fragment = CharacterResourceFragment.newInstance(character, ResourceType.EVENTS);
            getChildFragmentManager().beginTransaction().replace(R.id.eventsListContainer, fragment).commit();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @OnClick({R.id.detailButton, R.id.comiclinkButton, R.id.wikiButton})
    public void launchButtonLink(View button) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        MarvelUrl detailUrl = null;
        String type = (String)button.getTag();

        for (MarvelUrl marvelUrl : character.urls) {
            if (type.equals(marvelUrl.type)) {
                detailUrl = marvelUrl;
                break;
            }
        }

        if (detailUrl != null) {
            intent.setData(Uri.parse(detailUrl.url));
            startActivity(intent);
        } else {
            Toast.makeText(getContext(), getString(R.string.link_not_found), Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/kotlin/com/rcdvl/marvel/characterdetails/CharacterDetailsFragment.kt
package com.rcdvl.marvel.characterdetails

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.paginate.Paginate
import com.rcdvl.marvel.R
import com.rcdvl.marvel.characterlist.CharacterLoadingListItemCreator
import com.rcdvl.marvel.characterlist.CharactersAdapter
import com.rcdvl.marvel.networking.MarvelCharacter
import com.rcdvl.marvel.networking.MarvelListWrapper
import com.rcdvl.marvel.util.HorizontalSpaceItemDecoration
import kotlinx.android.synthetic.main.fragment_character_details.*
import java.util.*

/**
 * Created by renan on 3/18/16.
 */
class CharacterDetailsFragment : Fragment() {

    var character: MarvelCharacter? = null

    companion object {
        fun newInstance(character: MarvelCharacter): CharacterDetailsFragment {
            val fragment: CharacterDetailsFragment = CharacterDetailsFragment()
            var args = Bundle()
            args.putSerializable(CharactersAdapter.EXTRA_CHARACTER, character)
            fragment.arguments = args
            return fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        character = arguments.getSerializable(CharactersAdapter.EXTRA_CHARACTER) as MarvelCharacter?
        var view = inflater?.inflate(R.layout.fragment_character_details, container, false)
        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        Glide.with(context)
                .load(character?.thumbnail?.path + '.' + character?.thumbnail?.extension)
                .centerCrop()
                .crossFade()
                .into(characterImage);

        Glide.with(context)
                .load(R.drawable.background)
                .into(backgroundImage)

        characterName.text = character?.name

        if (TextUtils.isEmpty(character?.description)) {
            descriptionContainer.visibility = View.GONE
        } else {
            descriptionText.text = character?.description
        }

        comicsList.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        seriesList.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        storiesList.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        eventsList.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)

        if (character != null) {
            setupPagination(comicsList, "comics")
            setupPagination(seriesList, "series")
            setupPagination(storiesList, "stories")
            setupPagination(eventsList, "events")
        }

        comicsList.addItemDecoration(HorizontalSpaceItemDecoration())
        seriesList.addItemDecoration(HorizontalSpaceItemDecoration())
        storiesList.addItemDecoration(HorizontalSpaceItemDecoration())
        eventsList.addItemDecoration(HorizontalSpaceItemDecoration())

        hideIfEmpty(comicsContainer, character?.comics)
        hideIfEmpty(seriesContainer, character?.series)
        hideIfEmpty(storiesContainer, character?.stories)
        hideIfEmpty(eventsContainer, character?.events)

        setMiscButtonListener(detailButton, "detail")
        setMiscButtonListener(wikiButton, "wiki")
        setMiscButtonListener(comiclinkButton, "comiclink")

        contentLayout.requestFocus()

        super.onViewCreated(view, savedInstanceState)
    }

    private fun hideIfEmpty(view: View, list: MarvelListWrapper?) {
        if (list?.items == null || list?.items?.size == 0) view.visibility = View.GONE
    }

    private fun setupPagination(list: RecyclerView, resourceType: String) {
        var adapter: CharacterResourceAdapter? = CharacterResourceAdapter(ArrayList())
        list.adapter = adapter

        Paginate.with(list, Callback(resourceType, character!!.id, adapter!!))
                .setLoadingListItemCreator(CharacterLoadingListItemCreator())
                .build();
    }

    private fun setMiscButtonListener(button: View, type: String) {
        button.setOnClickListener({
            val intent = Intent(Intent.ACTION_VIEW)
            val detailUrl = character?.urls?.first { it.type.equals("$type", true) }

            if (detailUrl != null) {
                intent.data = Uri.parse(detailUrl.url)
                activity.startActivity(intent)
            }
        })

    }
}

